client/components/tinymce/plugins/simple-payments/dialog/index.jsx

```diff
@@ -181,11 +181,42 @@ class SimplePaymentsDialog extends Component {
 		return isNumber( this.props.editPaymentId );
 	}
 
+	checkUnsavedForm() {
+		const formIsUnsaved = this.state.activeTab === 'form' && this.props.formIsDirty;
+
+		if ( ! formIsUnsaved ) {
+			// No need to check anything, resolve to `accepted = true` right away.
+			return Promise.resolve( true );
+		}
+
+		// ask for confirmation
+		return new Promise( resolve => {
+			const { translate } = this.props;
+			accept(
+				translate( 'Wait! You have unsaved changes. Do you really want to discard them?' ),
+				resolve,
+				translate( 'Discard' ),
+				null,
+				{
+					isScary: true,
+				}
+			);
+		} );
+	}
+
+	handleDialogClose = () => {
+		// If there is a form that needs to be saved, ask for confirmation first.
+		// If not confirmed, the transition will be cancelled -- dialog remains opened.
+		this.checkUnsavedForm().then( accepted => accepted && this.props.onClose() );
+	};
+
 	handleChangeTabs = activeTab => {
 		if ( activeTab === 'form' ) {
 			this.showButtonForm( null );
 		} else {
```

Comment:
I know about the `no-else-return` rule, but I'm not always a big fan. It's great if we're returning early in case of error:
```js
function doSomething( param ) {
  if ( ! isValid( param ) ) {
    return null;
  }

  doTheMainJob();
  return result;
}
```
Here, `else` would just add one more indentation level and obscure the fact that one of the condition results is more important that other.

In our case though, there are two or more equally important branches. I think that explicit `else` is more appropriate here.